src/platform/GenericOTARequestorDriver.cpp

```diff
@@ -93,30 +152,162 @@ void GenericOTARequestorDriver::UpdateSuspended(System::Clock::Seconds32 delay)
         delay = kDefaultDelayedActionTime;
     }
 
-    ScheduleDelayedAction(UpdateFailureState::kAwaitingNextAction, delay,
-                          [](System::Layer *, void * context) { ToDriver(context)->mRequestor->ApplyUpdate(); });
+    ScheduleDelayedAction(
+        delay, [](System::Layer *, void * context) { ToDriver(context)->mRequestor->ApplyUpdate(); }, this);
 }
 
 void GenericOTARequestorDriver::UpdateDiscontinued()
 {
     VerifyOrDie(mImageProcessor != nullptr);
     mImageProcessor->Abort();
+
+    // Cancel all update timers
+    UpdateCancelled();
+
+    // Restart the periodic default provider timer
+    StartDefaultProviderTimer();
 }
 
+// Cancel all OTA update timers
 void GenericOTARequestorDriver::UpdateCancelled()
 {
-    // Platform might choose to schedule a retry here
+    // Cancel all OTA Update timers started by  OTARequestorDriver regardless of whether thery are running or not
+    CancelDelayedAction([](System::Layer *, void * context) { ToDriver(context)->mRequestor->DownloadUpdate(); }, this);
+    CancelDelayedAction(StartDelayTimerHandler, this);
+    CancelDelayedAction([](System::Layer *, void * context) { ToDriver(context)->mImageProcessor->Apply(); }, this);
+    CancelDelayedAction([](System::Layer *, void * context) { ToDriver(context)->mRequestor->ApplyUpdate(); }, this);
+}
+
+void GenericOTARequestorDriver::ScheduleDelayedAction(System::Clock::Seconds32 delay, System::TimerCompleteCallback action,
+                                                      void * aAppState)
+{
+    SystemLayer().StartTimer(std::chrono::duration_cast<System::Clock::Timeout>(delay), action, aAppState);
```

Comment:
This function used to take in an extra parameter `state` which is used to handle specific errors. Since that has been removed, it doesn't feel like this even needs to be a function? Callers of this function should just be able to call `SystemLayer().StartTimer` directly?

Reply:
It's a public method available to the OTARequestor. According to @tcarmelveilleux the OTA subsystem must abstract its timer system calls and potentially have platforms invoke  something other than StartTimer() here.

Reply:
Sure, except there is no case where OTARequestor needs to make this call now that all decision making is being done by the driver, correct?

Reply:
Yes, method is now private to the driver.